### ISSUE TITLE
`explicit_auto_deref` causes ` cannot return reference to function parameter`

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#15186]
Using the following flags

`--force-warn clippy::explicit_auto_deref`

this code:

```rust
fn foo(x: Box<&mut i32>) -> &mut i32 { &mut **x }
```

caused the following diagnostics:

```
    Checking _snippet_0 v0.1.0 (/tmp/icemaker_global_tempdir.IbV1PR0WiGNQ/icemaker_clippyfix_tempdir.1EXFQeNNggfj/_snippet_0)
warning: deref which would be done by auto-deref
 --> src/lib.rs:1:40
  |
1 | fn foo(x: Box<&mut i32>) -> &mut i32 { &mut **x }
  |                                        ^^^^^^^^ help: try: `&mut x`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref
  = note: requested on the command line with `--force-warn clippy::explicit-auto-deref`

warning: `_snippet_0` (lib) generated 1 warning (run `cargo clippy --fix --lib -p _snippet_0` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.11s

```

However after applying these diagnostics, the resulting code:

```rust
fn foo(mut x: Box<&mut i32>) -> &mut i32 { &mut x }
```
no longer compiled:

```
    Checking _snippet_0 v0.1.0 (/tmp/icemaker_global_tempdir.IbV1PR0WiGNQ/icemaker_clippyfix_tempdir.1EXFQeNNggfj/_snippet_0)
error[E0515]: cannot return reference to function parameter `x`
 --> src/lib.rs:1:44
  |
1 | fn foo(mut x: Box<&mut i32>) -> &mut i32 { &mut x }
  |                                            ^^^^^^ returns a reference to data owned by the current function

For more information about this error, try `rustc --explain E0515`.
error: could not compile `_snippet_0` (lib test) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `_snippet_0` (lib) due to 1 previous error

```

Version:
```
rustc 1.90.0-nightly (ed2d75978 2025-06-29)
binary: rustc
commit-hash: ed2d759783dc9de134bbb3f01085b1e6dbf539f3
commit-date: 2025-06-29
host: x86_64-unknown-linux-gnu
release: 1.90.0-nightly
LLVM version: 20.1.7

```
